packages/compass-shell/src/modules/worker-runtime/worker-shell-result.js

```diff
@@ -0,0 +1,116 @@
+  if (count === 0) {
+    return 'db.system.profile is empty.\nUse db.setProfilingLevel(2) will enable profiling.\nUse db.getCollection("system.profile").find() to show raw profile entries.';
+  }
+
+  // direct from old shell
+  const toret = result.map((x) => {
+    const res = `${x.op}    ${x.ns} ${x.millis}ms ${String(x.ts).substring(
+      0,
+      24
+    )}\n`;
+    let l = '';
+    for (const z in x) {
+      if (z === 'op' || z === 'ns' || z === 'millis' || z === 'ts') {
+        continue;
+      }
+
+      const val = x[z];
+      const mytype = typeof val;
+
+      if (mytype === 'object') {
+        l += z + ':' + inspect(val) + ' ';
+      } else if (mytype === 'boolean') {
+        l += z + ' ';
+      } else {
+        l += z + ':' + val + ' ';
+      }
+    }
+    return `${res}${l}`;
+  });
+
+  return toret.join('\n\n');
+}
+
+function isError(value) {
+  return typeof value.message === 'string' && typeof value.stack === 'string';
+}
+
+function isPrimitiveOrFunction(value) {
+  // any primitive type including 'null' and 'undefined',
+  // function and classes
+  return value !== Object(value) || typeof value === 'function';
+}
+
+export function serializeError(e) {
+  return { name: e.name, code: e.code, message: e.message, stack: e.stack };
+}
+
+// eslint-disable-next-line complexity
+export function serializeResult(result) {
+  const { type, printable } = result;
+
+  if (typeof printable === 'string' && type !== null) {
+    return result;
+    // return <SimpleTypeOutput value={printable} raw />;
+  }
+
+  if (isPrimitiveOrFunction(printable)) {
+    return { printable: inspect(printable) };
+    // return <SimpleTypeOutput value={printable} />;
+  }
+
+  if (type === 'Help') {
+    return result;
+    // return <HelpOutput value={value} />;
+  }
+
+  if (type === 'ShowDatabasesResult') {
+    return result;
+    // return <ShowDbsOutput value={value} />;
+  }
+
+  if (type === 'StatsResult') {
+    return result;
+    // return <StatsResultOutput value={value} />;
+  }
+
+  if (type === 'ListCommandsResult') {
+    return { printable: inspect(printable) };
+    // return <SimpleTypeOutput value={printable} />;
+  }
+
+  if (type === 'ShowCollectionsResult') {
+    return result;
+    // return <ShowCollectionsOutput value={printable} />;
+  }
+
+  if (type === 'Cursor') {
+    return { type, printable: printable.map(inspect) };
+    // return <CursorOutput value={printable} />;
+  }
+
+  if (type === 'CursorIterationResult') {
+    return { type, printable: printable.map(inspect) };
+    // return <CursorIterationResultOutput value={printable} />;
+  }
+
+  if (type === 'ShowProfileResult') {
+    return { printable: serializeProfileResult(printable) };
+    // return <ShowProfileOutput value={printable} />;
+  }
+
+  if (isError(printable)) {
+    return { printable: serializeError(printable) };
+    // return <ErrorOutput value={printable} />;
+  }
+
+  return { printable: inspect(printable) };
```

Comment:
Probably going to say something silly .. do you think `printable` here would be a bit overloaded with meanings? When consumed in the React side for some values will refer the `printable` property from the result of `ShellEvaluator.eval` and sometimes it would be the already rendered value from `inspect`, would it be expected  / understandable enough for people working on that part of the code?

Reply:
Oh, yeah, the interface can be much more explicit for the final implementation, I agree